File: intellij-lsp/src/com/github/gtache/actions/LSPQuickDocAction.scala
package com.github.gtache.actions

import com.github.gtache.editor.EditorEventManager
import com.intellij.codeInsight.documentation.actions.ShowQuickDocInfoAction
import com.intellij.openapi.actionSystem.{AnActionEvent, CommonDataKeys}
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.DumbAware

/**
  * Action overriding QuickDoc (CTRL+Q) : If editor / document is LSP-supported, shows the LSP doc, otherwise shows the default doc
  */
class LSPQuickDocAction extends ShowQuickDocInfoAction with DumbAware {
  private val LOG: Logger = Logger.getInstance(classOf[LSPQuickDocAction])

  override def actionPerformed(e: AnActionEvent): Unit = {
    val editor = e.getData(CommonDataKeys.EDITOR)
    EditorEventManager.forEditor(editor) match {
      case Some(manager) => manager.quickDoc(editor)
      case None => super.actionPerformed(e)
    }
  }
}

File: intellij-lsp/src/main/kotlin/com/github/gtache/lsp/actions/LSPQuickDocAction.kt
package com.github.gtache.lsp.actions

import com.github.gtache.lsp.editor.EditorEventManager
import com.github.gtache.lsp.settings.LSPState
import com.intellij.codeInsight.documentation.actions.ShowQuickDocInfoAction
import com.intellij.lang.LanguageDocumentation
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.fileTypes.PlainTextLanguage
import com.intellij.openapi.project.DumbAware
import com.intellij.psi.PsiManager

/**
 * Action overriding QuickDoc (CTRL+Q)
 */
class LSPQuickDocAction : ShowQuickDocInfoAction(), DumbAware {
    companion object {
        private val logger: Logger = Logger.getInstance(LSPQuickDocAction::class.java)
    }

    override fun actionPerformed(e: AnActionEvent): Unit {
        val editor = e.getData(CommonDataKeys.EDITOR)
        if (editor != null) {
            val file = FileDocumentManager.getInstance().getFile(editor.document)
            val project = editor.project
            if (file != null && project != null) {
                val language = PsiManager.getInstance(project).findFile(file)?.language
                //Hack for IntelliJ 2018 TODO proper way
                if (language != null && (LSPState.instance?.isAlwaysSendRequests != false ||
                            LanguageDocumentation.INSTANCE.allForLanguage(language).isEmpty() ||
                            (ApplicationInfo.getInstance().majorVersion.toInt() > 2017) && PlainTextLanguage.INSTANCE == language)) {
                    val manager = EditorEventManager.forEditor(editor)
                    if (manager != null) {
                        manager.quickDoc(editor)
                    } else super.actionPerformed(e)
                } else super.actionPerformed(e)
            } else super.actionPerformed(e)
        } else super.actionPerformed(e)
    }
}